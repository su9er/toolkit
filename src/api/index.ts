export * from './client'
export * from './schemaClient'
export * from './types'
