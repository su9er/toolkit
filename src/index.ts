export * from './middleware'
export * from './api'
