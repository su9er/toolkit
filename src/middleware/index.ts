export * from './onion'
export * from './compose'
