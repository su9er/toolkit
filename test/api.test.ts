import { describe, expect, it } from 'vitest'
import type { APISchema, RequestOptions } from '../src'
import { schemaClient } from '../src'

interface testModel {
  startTime: number
  endTime: number
}

interface TestAPISchema extends APISchema {
  'getTest': {
    request: void
    response: testModel
  }
  'setTest': {
    request: Partial<testModel>
    response: void
  }
  'download': {
    request: {
      key: string
    }
    response: {
      downloadUrl: string
    }
  }
  'getRes': {
    request: void
    response: Promise<string>
  }
}

describe('schema api client with plugins and configure', async() => {
  it('should return a configurable apis list', async() => {
    const configure = {
      getTest: 'GET /api/gettest',
      setTest: 'POST /api/settest',
      download: {
        // 支持参数占位符
        path: '/api/download/:id',
        method: 'POST',
        // 特殊接口请求
        headers: { 'x-download': 'xxx' },
      } as RequestOptions,
      // 使用配置函数
      getRes: () => {
        // get cache data
        const res = JSON.parse(window.localStorage.getItem('cache') || 'null')
        return Promise.resolve(res)
      },
    }
    const apis = schemaClient<TestAPISchema>(configure)
    expect(apis).toMatchInlineSnapshot(`
      {
        "download": [Function],
        "getRes": [Function],
        "getTest": [Function],
        "setTest": [Function],
      }
    `)
  })
})
