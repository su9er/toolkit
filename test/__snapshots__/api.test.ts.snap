// Vitest Snapshot v1

exports[`schema api client with plugins and configure > should return a configurable apis list 2`] = `
{
  "download": [Function],
  "getRes": [Function],
  "getTest": [Function],
  "setTest": [Function],
}
`;
